### PR TITLE
feat(auth): Support statically configured relays

### PR DESCRIPTION
**Pull request by RaduW**
This PR adds support for statically configuring downstream relays.

Previous to this PR endpoints that require Relay Authentication ( publickeys and projconfigs) needed to first get Relay information from upstream (ultimately from Sentry).

We would like to be able not to depened on ustream for validating relays and use a statically configured list of `(RelayId, PublicKey)` for authentication.

The new configuration is `<RELAY_CONFIG_DIR>/config.yml`
And has the following format:

```yaml
auth:
  ready: authenticated
  static_relays: # list of relay info objects
    "d33ccc9d-387b-4fdd-98cc-faeaa58fd39b": # relay ID
      public_key: TFxYsgWp0HtpVdwrlxS2ZfxL9g0R8OlK-sBoWrpvu3g # public key
      internal: true # true if the relay is internal
    "21fcbed2-1e5b-4fab-aafb-51ef025a2fa2":
      public_key: UDrH4Wl5tASA1QQGwQv4W08sq9m3hNQpYAUUgngQ-hc
      internal: false

```

### CODE REVIEW

File: relay-config/src/config.rs [hunk 1]
```diff
@@ -277,6 +277,38 @@ impl ConfigObject for Credentials {
...
+    /// The public key that this Relay uses to authenticate and sign requests.
+    pub public_key: PublicKey,
+
+    /// Marks an internal relay that has privileged access to more project configuration.
+    #[serde(default)]
+    pub internal: bool,
+}
+
+/// Contains a list of Relays that are statically configured (in a config file)
+/// For statically configured Relays there will be no upstream key lookup.
```
Comment (jan-auer) [inline relay-config/src/config.rs]:
```suggestion
/// Contains a list of Relays that are statically configured in a config file.
///
/// For statically configured Relays there will be no upstream key lookup.
```

File: relay-config/src/config.rs [hunk 2]
```diff
@@ -277,6 +277,38 @@ impl ConfigObject for Credentials {
...
+#[derive(Debug, Serialize, Deserialize, Clone, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct StaticRelays {
+    /// a list of Relay that will be resolved statically
```
Comment (jan-auer) [inline relay-config/src/config.rs]:
```suggestion
    /// A list of Relays that will be resolved statically.
    ///
    /// Relay IDs must be unique within this list. Duplicate entries lead to undefined behavior.
```

File: relay-config/src/config.rs [hunk 3]
```diff
@@ -895,9 +928,15 @@ impl Config {
 
         let config = Config {
             values: ConfigValues::load(&path)?,
-            credentials: match fs::metadata(Credentials::path(&path)) {
-                Ok(_) => Some(Credentials::load(&path)?),
-                Err(_) => None,
+            credentials: if Credentials::path(&path).exists() {
```
Comment (jan-auer) [inline relay-config/src/config.rs:984]: Please revert to using `fs::metadata` instead of `exists`, since it also checks for access permissions.

File: relay-config/src/config.rs [hunk 4]
```diff
@@ -277,6 +277,38 @@ impl ConfigObject for Credentials {
     }
 }
 
+/// Information on a downstream Relay.
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct StaticRelayInfo {
```
Comment (jan-auer) [inline relay-config/src/config.rs]: Ideally we manage to share the pre-existing `RelayInfo` type instead of creating a second type. That said, RelayInfo from the API does not contain the `id`, so let's stick with a separate type for now.

File: relay-config/src/config.rs [hunk 5]
```diff
@@ -277,6 +277,38 @@ impl ConfigObject for Credentials {
...
+    /// the id of the relay
```
Comment (jan-auer) [inline relay-config/src/config.rs]:
```suggestion
    /// The identifier of the Relay included in every signed request.
```

File: relay-server/src/extractors/signed_json.rs [hunk 6]
```diff
@@ -37,24 +37,26 @@ impl ResponseError for SignatureError {
...
-                    .get($name)
-                    .ok_or(SignatureError::MissingHeader($name))
-                    .and_then(|value| value
-                        .to_str()
-                        .map_err(|_| SignatureError::MalformedHeader($name))))
-            };
-        }
-
-        let relay_id: RelayId = tryf!(extract_header!("X-Sentry-Relay-Id")
+        let relay_id: RelayId = tryf!(tryf!(extract_header(req, "X-Sentry-Relay-Id"))
```
Comment (jan-auer) [inline relay-server/src/extractors/signed_json.rs]: This is an unrelated change without functional impact. Can we back this out into a separate PR, please?

File: tests/integration/fixtures/relay.py [hunk 7]
```diff
@@ -44,7 +46,12 @@ def shutdown(self, sig=signal.SIGKILL):
 @pytest.fixture
 def relay(mini_sentry, random_port, background_process, config_dir):
     def inner(
-        upstream, options=None, prepare=None, external=None, wait_healthcheck=True
+        upstream,
+        options=None,
+        prepare=None,
+        external=None,
+        wait_healthcheck=True,
+        static_relays_config=None,
```
Comment (jan-auer) [inline tests/integration/fixtures/relay.py]: Let's remove reundant suffices.

```suggestion
        static_relays=None,
```

File: requirements-test.txt [hunk 8]
```diff
@@ -9,3 +9,4 @@ redis==3.4.1
 requests==2.23.0
 git+git://github.com/getsentry/sentry-python@e234998ae82a9cffa6fb3718801c55ba24a86bab#egg=sentry_sdk
 werkzeug==0.15.5
+sentry-relay==0.8.5
```
Comment (jan-auer) [inline requirements-test.txt:12]: Please do not include librelay in Relay binary integration tests. This forces another (release!) build of the python library, which increases test times, and unnecessarily increases the test surface. Please hard code the expectations, instead.

File: relay-config/src/config.rs [hunk 9]
```diff
@@ -277,6 +277,38 @@ impl ConfigObject for Credentials {
...
+    pub id: RelayId,
+
+    /// The public key that this Relay uses to authenticate and sign requests.
+    pub public_key: PublicKey,
+
+    /// Marks an internal relay that has privileged access to more project configuration.
+    #[serde(default)]
+    pub internal: bool,
```
Comment (jan-auer) [inline relay-config/src/config.rs:289]: At some point we had discussed that these Relays are trusted by definition, similar to how this was handled in Sentry. Did you find new reasons to include this flag?

**Review by jan-auer — CHANGES_REQUESTED** [review]

File: relay-config/src/config.rs (hunk 4 above)
Comment (jan-auer) [reply relay-config/src/config.rs]: Follow-up after reviewing https://github.com/getsentry/sentry/pull/26155: Since we use a mapping by Relay-ID there, I propose the following:

- Move `relay_server::actors::RelayInfo` here, and remove the `StaticRelayInfo` type.
- Change the config file to be `HashMap<RelayId, RelayInfo>` or `BTreeMap<RelayId, RelayInfo>`.
- Update `RelayCache` constructor to clone the mapping rather than recollect it.

File: relay-config/src/config.rs (hunk 9 above)
Comment (RaduW) [reply relay-config/src/config.rs:289]: Yes, we did discuss but I see no reason/downside to limit it to trusted relays. 
The reasons I implemented it like that are:
* makes it clear from the configuration how the relays are treated
* makes it possible for us ( or more probable a large enterprise) to chain its own relays without having the downstream relays depend on registration). This would enable the enterprise to handle a similar scenario to what we are building the  feature for( the org to bring up additional downstream relays during an incident).

So giving the above and the fact that it does not complicate the handling in any way (we create a Relay object and we have to give it an internal flag, rather than hard coding it we take it from config) I thought it would be worth it having it in the configuration,

The only downside I can think of is that we need another entry (the internal flag) in the configuration file. The field has a default ( which is false, for safety) so although it can be omitted in our use case we need overwritten  to `true` .

File: relay-config/src/config.rs (hunk 3 above)
Comment (RaduW) [reply relay-config/src/config.rs:984]: It just delegates to `fs::metadata()` so it does not do less (or more than it).
I think it makes the code clearer but if you want I can change it back.

```
impl Path {
    ...
    pub fn exists(&self) -> bool {
        fs::metadata(self).is_ok()
    }
```

File: relay-server/src/extractors/signed_json.rs (hunk 6 above)
Comment (RaduW) [reply relay-server/src/extractors/signed_json.rs]: I had a suspicion you are going to say that :-) , placed it in a separate commit (I'll pick it out).

File: tests/integration/fixtures/relay.py (hunk 7 above)
Comment (RaduW) [reply tests/integration/fixtures/relay.py]: Since we are constructing a relay object and passing a mini_sentry, I wanted to make it clear that we don't pass relay objects but configs, i think that in this case the naming is not redundant but I can change it if you prefer.

File: relay-config/src/config.rs (hunk 4 above)
Comment (RaduW) [reply relay-config/src/config.rs]: Ok, will do, I explicitly made it a Vector and convert it in RelayCache because I think it is a lot nicer to have a list in the configuration file than a dictionary (where it is not very explicit what the key represents).

File: requirements-test.txt (hunk 8 above)
Comment (RaduW) [reply requirements-test.txt:12]: looking for a solution...

Comment (RaduW) [reply requirements-test.txt:12]: After discussing the alternatives with @untitaker we came to the conclusion that the most reasonable approach is to leave the sentry-relay dependency and add a dependabot task that will check for the dependencies on a few packages in the `requirements-test.txt` ( sentry-relay, sentry-python, pytest-sentry)

File: relay-server/src/actors/relays.rs [hunk 10]
```diff
@@ -108,16 +108,19 @@ pub struct RelayCache {
     config: Arc<Config>,
     upstream: Addr<UpstreamRelay>,
     relays: HashMap<RelayId, RelayState>,
+    static_relays: HashMap<RelayId, RelayInfo>,
     relay_channels: HashMap<RelayId, RelayInfoChannel>,
 }
 
 impl RelayCache {
     pub fn new(config: Arc<Config>, upstream: Addr<UpstreamRelay>) -> Self {
+        let static_relays = config.static_relays().relays.clone();
```
Comment (jan-auer) [inline relay-server/src/actors/relays.rs]: _nit_: This could be inlined.

**Review by jan-auer — APPROVED** [review]
Thanks, good to go!

**Review by jan-auer — APPROVED** [review]
